flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileSourceSplitReader.java

```diff
@@ -0,0 +1,110 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.file.src.impl;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitReader;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsAddition;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsChange;
+import org.apache.flink.connector.file.src.FileSourceSplit;
+import org.apache.flink.connector.file.src.reader.BulkFormat;
+import org.apache.flink.connector.file.src.util.RecordAndPosition;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * The {@link SplitReader} implementation for the file source.
+ */
+final class FileSourceSplitReader<T> implements SplitReader<RecordAndPosition<T>, FileSourceSplit> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(FileSourceSplitReader.class);
+
+	private final Configuration config;
+	private final BulkFormat<T> readerFactory;
+
+	private final Queue<FileSourceSplit> splits;
+
+	@Nullable
+	private BulkFormat.Reader<T> currentReader;
+	@Nullable
+	private String currentSplitId;
+
+	public FileSourceSplitReader(Configuration config, BulkFormat<T> readerFactory) {
+		this.config = config;
+		this.readerFactory = readerFactory;
+		this.splits = new ArrayDeque<>();
+	}
+
+	@Override
+	public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
+		checkSplitOrStartNext();
+
+		final BulkFormat.RecordIterator<T> nextBatch = currentReader.readBatch();
+		return nextBatch == null ? finishSplit() : FileRecords.forRecords(currentSplitId, nextBatch);
+	}
+
+	@Override
+	public void handleSplitsChanges(final SplitsChange<FileSourceSplit> splitChange) {
+		if (!(splitChange instanceof SplitsAddition)) {
+			throw new UnsupportedOperationException(String.format(
+					"The SplitChange type of %s is not supported.", splitChange.getClass()));
+		}
+
+		LOG.debug("Handling split change {}", splitChange);
+		splits.addAll(splitChange.splits());
+	}
+
+	@Override
+	public void wakeUp() {}
```

Comment:
This is a good question. TBH, I am not sure how to implement waking up here, or if we need it.

  - **Do we need wakeup?** I think not necessarily, because this method only stalls for a limited amount of time, until the I/O stream has delivered the next records. There is guaranteed to be something coming soon. Not waking up means some other operation (split changes) may get processed a second later.

  - **Can we implement wakeup?** I would not know how. Many I/O streams and decoders or parsers are not well suited for interruptions (don't react or become corrupt) and there is no common way to interrupt an I/O request.

  - **How can we speed up cancallation?** The only case where I am worried about a speedy response is cancellation. But we can speed this up by closing the reader directly. Closing reader/stream means that any reader on that stream will immediately exit with an IOException. Which is fine, because we ignore exceptions once cancellation has been initiated.

Reply:
You are right. I agree the current solution is a good default behavior. It is tricky because we don't know whether `Reader#readBatch()` / `Reader#read()` method can be waken up. Maybe we can add a default wakeup method in the `Reader` interface. So for some of the implementation who can be waken up, they would behave more elegantly.

Reply:
Yeah, I don't have a strong use case at this point. I was thinking about some reader implementation using Java async IO with a selector. But that is not really needed here because it is OK for `readBatch()` / `read()` to be blocking.

Reply:
Java async IO with a selector would probably not be a file format but a different Source with a different Split Reader. There we have the wakeup() method already.